File: apps/web/lib/api/workspace-assistant.ts
import { apiRequest } from "@/lib/api/client";

export type AssistantScopeType =
  | "tenant"
  | "client"
  | "matter"
  | "ip_docket"
  | "ip_asset"
  | "trademark_application"
  | "ip_proceeding"
  | "matter_document"
  | "ip_document";

export type AssistantScopeInput = {
  scope_type: AssistantScopeType;
  scope_id: string;
};

export type AssistantScopeOption = AssistantScopeInput & {
  label: string;
  secondary_text: string | null;
  href: string;
  resource_version: string;
};

export type AssistantScopeRecord = AssistantScopeInput & {
  resource_version: string | null;
  ordinal: number;
};

export type AssistantSessionRecord = {
  id: string;
  title: string;
  status: "active" | "archived";
  version: number;
  policy_version: number;
  scope_state: "current" | "permission_changed";
  scopes: AssistantScopeRecord[];
  retention_expires_at: string;
  archived_at: string | null;
  created_at: string;
  updated_at: string;
};

export type AssistantSessionSummary = Pick<
  AssistantSessionRecord,
  | "id"
  | "title"
  | "status"
  | "version"
  | "retention_expires_at"
  | "archived_at"
  | "created_at"
  | "updated_at"
>;

export type AssistantCitation = {
  id: string;
  ordinal: number;
  source_type: string;
  source_id: string;
  source_version: string;
  source_sha256: string | null;
  source_url: string | null;
  label: string;
  excerpt: string | null;
  verified_at: string | null;
};

export type AssistantModelMetadata = {
  run_id: string;
  provider: string;
  model: string;
  purpose: string;
  prompt_tokens: number;
  completion_tokens: number;
  latency_ms: number;
  status: string;
};

export type AssistantProposedAction = {
  proposal_id: string;
  action_type: "navigation" | "search" | "draft" | "task" | "field_update";
  label: string;
  href: string | null;
  target_type: string | null;
  target_id: string | null;
  instruction: string | null;
  requires_confirmation: boolean;
  execution_available: boolean;
};

export type AssistantTurn = {
  id: string;
  sequence: number;
  role: "user" | "assistant";
  status: "queued" | "completed" | "abstained" | "failed" | "cancelled";
  render_status: "visible" | "permission_changed";
  content: string;
  citations: AssistantCitation[];
  model: AssistantModelMetadata | null;
  suggested_searches: string[];
  proposed_actions: AssistantProposedAction[];
  created_at: string;
};

export type AssistantAskResponse = {
  session: AssistantSessionRecord;
  user_turn: AssistantTurn;
  assistant_turn: AssistantTurn;
};

export type AssistantSessionExport = {
  schema_version: number;
  exported_at: string;
  session: AssistantSessionRecord;
  turns: AssistantTurn[];
  retention_disposition: string;
};

export function searchAssistantScopes(query: string, limit = 12) {
  const params = new URLSearchParams({ q: query.trim(), limit: String(limit) });
  return apiRequest<{ query: string; items: AssistantScopeOption[]; truncated: boolean }>(
    `/api/workspace-assistant/scope-options?${params}`,
    { timeoutMs: 15_000 },
  );
}

export function listAssistantSessions(status?: "active" | "archived") {
  const params = new URLSearchParams({ limit: "25", offset: "0" });
  if (status) params.set("status", status);
  return apiRequest<{
    items: AssistantSessionSummary[];
    limit: number;
    offset: number;
    has_more: boolean;
  }>(`/api/workspace-assistant/sessions?${params}`);
}

export function createAssistantSession(title: string, scopes: AssistantScopeInput[]) {
  return apiRequest<AssistantSessionRecord>("/api/workspace-assistant/sessions", {
    method: "POST",
    body: { title, scopes },
  });
}

export function getAssistantSession(sessionId: string) {
  return apiRequest<AssistantSessionRecord>(
    `/api/workspace-assistant/sessions/${sessionId}`,
  );
}

export function replaceAssistantScopes(
  sessionId: string,
  expectedVersion: number,
  scopes: AssistantScopeInput[],
) {
  return apiRequest<AssistantSessionRecord>(
    `/api/workspace-assistant/sessions/${sessionId}/scopes`,
    { method: "PUT", body: { expected_version: expectedVersion, scopes } },
  );
}

export function archiveAssistantSession(sessionId: string, expectedVersion: number) {
  return apiRequest<AssistantSessionRecord>(
    `/api/workspace-assistant/sessions/${sessionId}/archive`,
    { method: "POST", body: { expected_version: expectedVersion } },
  );
}

export function askWorkspaceAssistant(
  sessionId: string,
  expectedVersion: number,
  question: string,
) {
  return apiRequest<AssistantAskResponse>(
    `/api/workspace-assistant/sessions/${sessionId}/ask`,
    {
      method: "POST",
      body: { expected_version: expectedVersion, question },
      timeoutMs: 75_000,
    },
  );
}

export function listAssistantTurns(sessionId: string) {
  return apiRequest<{ items: AssistantTurn[]; limit: number; offset: number; has_more: boolean }>(
    `/api/workspace-assistant/sessions/${sessionId}/turns?limit=50&offset=0`,
  );
}

export function exportAssistantSession(sessionId: string) {
  return apiRequest<AssistantSessionExport>(
    `/api/workspace-assistant/sessions/${sessionId}/export`,
  );
}

export function deleteAssistantSession(sessionId: string) {
  return apiRequest<void>(`/api/workspace-assistant/sessions/${sessionId}`, {
    method: "DELETE",
  });
}

File: apps/api/src/caseops_api/schemas/workspace_assistant.py
from __future__ import annotations

from datetime import date, datetime
from typing import Literal

from pydantic import BaseModel, Field, field_validator

AssistantScopeType = Literal[
    "tenant",
    "client",
    "matter",
    "ip_docket",
    "ip_asset",
    "trademark_application",
    "ip_proceeding",
    "matter_document",
    "ip_document",
]


class AssistantScopeInput(BaseModel):
    scope_type: AssistantScopeType
    scope_id: str = Field(min_length=1, max_length=36)

    @field_validator("scope_id")
    @classmethod
    def normalize_scope_id(cls, value: str) -> str:
        normalized = value.strip()
        if not normalized:
            raise ValueError("scope_id cannot be blank")
        return normalized


class AssistantSessionCreateRequest(BaseModel):
    title: str = Field(min_length=1, max_length=160)
    scopes: list[AssistantScopeInput] = Field(min_length=1, max_length=24)

    @field_validator("title")
    @classmethod
    def normalize_title(cls, value: str) -> str:
        normalized = " ".join(value.split())
        if not normalized:
            raise ValueError("title cannot be blank")
        return normalized


class AssistantScopeReplaceRequest(BaseModel):
    expected_version: int = Field(ge=1)
    scopes: list[AssistantScopeInput] = Field(min_length=1, max_length=24)


class AssistantSessionArchiveRequest(BaseModel):
    expected_version: int = Field(ge=1)


class AssistantScopeRecord(BaseModel):
    scope_type: AssistantScopeType
    scope_id: str
    resource_version: str | None
    ordinal: int


class AssistantSessionRecord(BaseModel):
    id: str
    title: str
    status: Literal["active", "archived"]
    version: int
    policy_version: int
    scope_state: Literal["current", "permission_changed"]
    scopes: list[AssistantScopeRecord]
    retention_expires_at: datetime
    archived_at: datetime | None
    created_at: datetime
    updated_at: datetime


class AssistantSessionSummary(BaseModel):
    id: str
    title: str
    status: Literal["active", "archived"]
    version: int
    retention_expires_at: datetime
    archived_at: datetime | None
    created_at: datetime
    updated_at: datetime


class AssistantSessionListResponse(BaseModel):
    items: list[AssistantSessionSummary]
    limit: int
    offset: int
    has_more: bool


class AssistantScopeOption(BaseModel):
    scope_type: AssistantScopeType
    scope_id: str
    label: str
    secondary_text: str | None
    href: str
    resource_version: str


class AssistantScopeSearchResponse(BaseModel):
    query: str
    items: list[AssistantScopeOption]
    truncated: bool


class AssistantAskRequest(BaseModel):
    expected_version: int = Field(ge=1)
    question: str = Field(min_length=2, max_length=2000)

    @field_validator("question")
    @classmethod
    def normalize_question(cls, value: str) -> str:
        normalized = " ".join(value.split())
        if len(normalized) < 2:
            raise ValueError("question must contain at least two characters")
        return normalized


class AssistantCitationRecord(BaseModel):
    id: str
    ordinal: int
    source_type: str
    source_id: str
    source_version: str
    source_sha256: str | None
    source_url: str | None
    label: str
    excerpt: str | None
    verified_at: datetime | None


class AssistantCitationOpenResponse(BaseModel):
    citation_id: str
    source_url: str


class AssistantModelMetadata(BaseModel):
    run_id: str
    provider: str
    model: str
    purpose: str
    prompt_tokens: int
    completion_tokens: int
    latency_ms: int
    status: str


class AssistantProposedAction(BaseModel):
    proposal_id: str
    action_type: Literal["navigation", "search", "draft", "task", "field_update"]
    label: str
    href: str | None = None
    target_type: str | None = None
    target_id: str | None = None
    target_version: str | None = None
    target_label: str | None = None
    instruction: str | None = None
    requires_confirmation: bool
    execution_available: bool = False
    unavailable_reason: str | None = None


AssistantActionType = Literal["draft", "task", "field_update"]
AssistantMatterField = Literal[
    "title",
    "description",
    "matter_type",
    "client_name",
    "opposing_party",
    "opposing_counsel",
    "practice_area",
    "court_name",
    "judge_name",
]


class AssistantActionInput(BaseModel):
    model_config = {"extra": "forbid"}

    title: str | None = Field(default=None, max_length=255)
    description: str | None = Field(default=None, max_length=4000)
    due_on: date | None = None
    priority: Literal["low", "medium", "high", "urgent"] = "medium"
    draft_type: Literal["brief", "notice", "reply", "memo", "other"] = "memo"
    template_key: str | None = Field(default=None, min_length=3, max_length=60)
    field_name: AssistantMatterField | None = None
    field_value: str | None = Field(default=None, max_length=4000)


class AssistantActionPreviewRequest(BaseModel):
    expected_version: int = Field(ge=1)
    turn_id: str = Field(min_length=1, max_length=36)
    proposal_id: str = Field(min_length=1, max_length=64)
    input: AssistantActionInput


class AssistantActionConfirmRequest(BaseModel):
    expected_version: int = Field(ge=1)
    preview_token: str = Field(min_length=64, max_length=64)


class AssistantActionChangeRecord(BaseModel):
    field: str
    before: str | None
    after: str | None


class AssistantActionPreviewResponse(BaseModel):
    preview_id: str
    proposal_id: str
    action_type: AssistantActionType
    status: Literal["pending", "superseded", "confirmed"]
    session_version: int
    resulting_session_version: int | None = None
    target_type: str
    target_id: str
    target_label: str
    summary: str
    changes: list[AssistantActionChangeRecord]
    warnings: list[str]
    required_capabilities: list[str]
    preview_token: str
    expires_at: datetime
    result_type: str | None = None
    result_id: str | None = None
    result_href: str | None = None


class AssistantTurnRecord(BaseModel):
    id: str
    sequence: int
    role: Literal["user", "assistant"]
    status: Literal["queued", "completed", "abstained", "failed", "cancelled"]
    render_status: Literal["visible", "permission_changed"]
    content: str
    citations: list[AssistantCitationRecord]
    model: AssistantModelMetadata | None
    suggested_searches: list[str]
    proposed_actions: list[AssistantProposedAction]
    created_at: datetime


class AssistantAskResponse(BaseModel):
    session: AssistantSessionRecord
    user_turn: AssistantTurnRecord
    assistant_turn: AssistantTurnRecord


class AssistantTurnListResponse(BaseModel):
    items: list[AssistantTurnRecord]
    limit: int
    offset: int
    has_more: bool


class AssistantSessionExportResponse(BaseModel):
    schema_version: int
    exported_at: datetime
    session: AssistantSessionRecord
    turns: list[AssistantTurnRecord]
    retention_disposition: str
